icu4j/main/core/src/main/java/com/ibm/icu/text/UnicodeSet.java

```diff
@@ -5127,5 +5140,237 @@ public static void setDefaultXSymbolTable(XSymbolTable xSymbolTable) {
+         * Note "exact count of the number of elements that would be encountered"
+         */
+        private final static int CHARACTERISTICS = Spliterator.ORDERED | Spliterator.DISTINCT | Spliterator.NONNULL;
+
+        private final UnicodeSet unicodeSet;
+        private final CodePointsIteratorInt cpi;
+
+        CodePointsSpliterator(UnicodeSet unicodeSet) {
+            this.unicodeSet = unicodeSet;
+            cpi = new CodePointsIteratorInt(unicodeSet);
+        }
+
+        @Override
+        public long estimateSize() {
+            return unicodeSet.size() - unicodeSet.strings.size();
+        }
+
+        @Override
+        public int characteristics() {
+            return unicodeSet.isFrozen() ? Spliterator.IMMUTABLE | CHARACTERISTICS : CHARACTERISTICS;
+        }
+
+        @Override
+        public Spliterator.OfInt trySplit() {
+            /* From the doc:
+             *   > This method may return null for any reason, including emptiness, inability to split after
+             *   > traversal has commenced, data structure constraints, and efficiency considerations.
+             */
+            return null;
+        }
+
+        @Override
+        public boolean tryAdvance(IntConsumer action) {
+            if (action == null)
+                throw new NullPointerException();
+
+            if (cpi.hasNext()) {
+                action.accept(cpi.next());
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /* This class is optimized to iterate on code points and will be used to implement both
+     * the Iterator<Integer> (Integer, boxed value) and the Spliterator.OfInt (int primitive).
+     * It looks exactly like an Iterator<Integer>, but works on the primitive int,
+     * so it can't implement Iterator.
+     */
+    static private class CodePointsIteratorInt {
+        private final UnicodeSet unicodeSet;
+        private final int lastRange; 
+        private int currentRange = 0;
+        private int startRange;
+        private int endRange;
+
+        public CodePointsIteratorInt(UnicodeSet unicodeSet) {
+            this.unicodeSet = unicodeSet;
+            lastRange = unicodeSet.len - 1;
+            currentRange = 0;
+            startRange = unicodeSet.list[currentRange++];
+            if (lastRange > 0) { // not an empty list
+                endRange = unicodeSet.list[currentRange++];
+            } else {
+                endRange = startRange; // should be HIGH, the guard value
+            }
+        }
+
+        public boolean hasNext() {
+            if (startRange < endRange) {
+                // We can still iterate on code points in the current range
+                return true;
+            }
+            return currentRange < lastRange;
+        }
+
+        public int next() {
+            if (startRange >= endRange) {
+                startRange = unicodeSet.list[currentRange++];
+                endRange = unicodeSet.list[currentRange++];
```

Comment:
Done.

And added tests.
It didn't happen because it was protected by `hasNext()`
And since the correct use is `while (hasNext() ) { ...next() }` it was all good.